client/src/main/scala/org/http4s/client/PoolManager.scala

```diff
@@ -85,30 +107,37 @@ private final class PoolManager[F[_], A <: Connection[F]](
         if (!isClosed) {
           @tailrec
           def go(): Unit =
-            idleQueue.dequeueFirst(_.requestKey == key) match {
+            getConnectionFromQueue(key) match {
               case Some(conn) if !conn.isClosed =>
                 logger.debug(s"Recycling connection: $stats")
                 callback(Right(NextConnection(conn, fresh = false)))
 
               case Some(closedConn) =>
                 logger.debug(s"Evicting closed connection: $stats")
-                allocated -= 1
+                decrConnection(key)
                 go()
 
-              case None if allocated < maxTotal =>
+              case None
+                  if curTotal < maxTotal && allocated.getOrElse(key, 0) < maxConnectionsPerRequestKey(
+                    key) =>
                 logger.debug(s"Active connection not found. Creating new one. $stats")
                 createConnection(key, callback)
 
-              case None if idleQueue.nonEmpty =>
+              case None if idleQueue.get(key).exists(_.nonEmpty) =>
```

Comment:
Why do we prefer to open a new connection to taking one from the `idleQueue`?

Reply:
Not sure, kept the same logic as the previous code here.